File: backend_flask/services/user_session_service.py
from config.db import execute_write, fetch_one

_schema_checked = False


def ensure_user_session_schema():
    global _schema_checked

    if _schema_checked:
        return

    row = fetch_one(
        """
        SELECT COLUMN_NAME
        FROM INFORMATION_SCHEMA.COLUMNS
        WHERE TABLE_SCHEMA = DATABASE()
          AND TABLE_NAME = 'users'
          AND COLUMN_NAME = 'active_session_id'
        LIMIT 1
        """,
    )

    if not row:
        execute_write(
            """
            ALTER TABLE users
            ADD COLUMN active_session_id VARCHAR(64) NULL
            """,
        )

    _schema_checked = True

File: backend/src/services/userSession.service.js
import { pool } from "../config/db.js";

let userSessionSchemaPromise = null;

async function ensureUserSessionSchemaInternal() {
  const [rows] = await pool.execute(
    `SELECT COLUMN_NAME
     FROM INFORMATION_SCHEMA.COLUMNS
     WHERE TABLE_SCHEMA = DATABASE()
       AND TABLE_NAME = 'users'
       AND COLUMN_NAME = 'active_session_id'
     LIMIT 1`,
  );

  if (!rows.length) {
    await pool.execute(
      `ALTER TABLE users
       ADD COLUMN active_session_id VARCHAR(64) NULL AFTER banned_reason`,
    );
  }
}

export async function ensureUserSessionSchema() {
  if (!userSessionSchemaPromise) {
    userSessionSchemaPromise = ensureUserSessionSchemaInternal().catch(
      (error) => {
        userSessionSchemaPromise = null;
        throw error;
      },
    );
  }

  return userSessionSchemaPromise;
}
